### PR TITLE
[PseudoProbe] Make probe discriminator compatible with dwarf base discriminator

### PR DESCRIPTION
**Pull request by wlei-llvm**
It's useful if the probe-based build can consume a dwarf based profile(e.g. the profile transition), before there is a conflict for the discriminator, this change tries to mitigate the issue by encoding the dwarf base discriminator into the probe discriminator. 
As the num of probe id(num of basic block and calls) starts from 1, there are some unused space. We try to reuse some bit of the probe id.  
The new encode rule is:
- Use a bit to [28:28] to indicate whether dwarf base discriminator is encoded.(fortunately we can borrow this bit from the `PseudoProbeType`)
- If the bit is set, use [15:3] for probe id, [18:16] for dwarf base discriminator. Otherwise, still use [18:3] for probe id.

Note that these doesn't affect the original probe id capacity, we still prioritize probe id encoding, i.e. the base discriminator is not encoded when probe id is bigger than [15:3].
 
Then adjust `getBaseDiscriminatorFromDiscriminator` to use the  base discriminator from the probe discriminator.

### CODE REVIEW

File: llvm/include/llvm/IR/PseudoProbe.h [hunk 1]
```diff
@@ -44,26 +44,52 @@ struct PseudoProbeDwarfDiscriminator {
...
 
+  static uint32_t extractDwarfBaseDiscriminator(uint32_t Value) {
+    if (isDwarfBaseDiscriminatorEncoded(Value))
+      return (Value >> 16) & 0x7;
+    // Return an invalid value to indicate the dwarf base discriminator is not
+    // encoded.
+    return 0xFFFFFFFF;
+  }
+
+  static uint32_t isDwarfBaseDiscriminatorEncoded(uint32_t Value) {
```
Comment (WenleiHe) [inline llvm/include/llvm/IR/PseudoProbe.h:86]: Given this is `is...`, should this return a boolean, and just check `Value & mask`?

File: llvm/include/llvm/IR/PseudoProbe.h [hunk 2]
```diff
@@ -44,26 +44,52 @@ struct PseudoProbeDwarfDiscriminator {
...
+      return (Value >> 3) & 0x1FFF;
     return (Value >> 3) & 0xFFFF;
   }
 
+  static uint32_t extractDwarfBaseDiscriminator(uint32_t Value) {
+    if (isDwarfBaseDiscriminatorEncoded(Value))
+      return (Value >> 16) & 0x7;
+    // Return an invalid value to indicate the dwarf base discriminator is not
+    // encoded.
+    return 0xFFFFFFFF;
```
Comment (WenleiHe) [inline llvm/include/llvm/IR/PseudoProbe.h]: Should we always expect `isDwarfBaseDiscriminatorEncoded` be checked before calling `extractDwarfBaseDiscriminator`? Basically turning this into an assertion.

Comment (wlei-llvm) [reply llvm/include/llvm/IR/PseudoProbe.h]: I just changed the return type to `optional`, it's required to check in the caller, then we don't need the special `0xFFFFFFFF`.

File: llvm/include/llvm/IR/PseudoProbe.h (hunk 1 above)
Comment (wlei-llvm) [reply llvm/include/llvm/IR/PseudoProbe.h]: Good catch, fixed.

**Review by WenleiHe — COMMENTED** [review]
Looks good. It would be good to test:

1) This change has zero impact on CSSPGO builds (both pass1 and pass2). 
2) Generally how many (%) of total dwarf discriminator can fit in the 3 bits.

**Review by WenleiHe — APPROVED** [review]
lgtm, thanks.

**Comment by wlei-llvm** [comment]
> 1. This change has zero impact on CSSPGO builds (both pass1 and pass2).
Tested on our internal services, both pass1 and pass2 are good.

> 2. Generally how many (%) of total dwarf discriminator can fit in the 3 bits.

Dumped the stats of this. So the the  3 bits.( <= 7)  discriminator account for roughly 96% of the total discriminator. 
Actually most are the zero discriminator about 70%,  value >=1 && value <=7 acount about 25%.

**Pull request merged**